contracts/utils/Swap.sol

```diff
@@ -0,0 +1,454 @@
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "../interfaces/ISwap.sol";
+import {IERC20Detailed} from "../interfaces/IERC20Detailed.sol";
+import {SafeMath} from "@openzeppelin/contracts/utils/math/SafeMath.sol";
+
+contract Swap is ISwap, Ownable {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    bytes32 public constant DOMAIN_TYPEHASH =
+        keccak256(
+            abi.encodePacked(
+                "EIP712Domain(",
+                "string name,",
+                "string version,",
+                "uint256 chainId,",
+                "address verifyingContract",
+                ")"
+            )
+        );
+
+    bytes32 public constant BID_TYPEHASH =
+        keccak256(
+            abi.encodePacked(
+                "Bid(",
+                "uint256 swapId,",
+                "uint256 nonce,",
+                "address signerWallet,",
+                "uint256 sellAmount,",
+                "uint256 buyAmount,",
+                "address referrer",
+                ")"
+            )
+        );
+
+    bytes32 public constant DOMAIN_NAME = keccak256("RIBBON SWAP");
+    bytes32 public constant DOMAIN_VERSION = keccak256("1");
+    uint256 public immutable DOMAIN_CHAIN_ID;
+    bytes32 public immutable DOMAIN_SEPARATOR;
+
+    uint256 internal constant MAX_PERCENTAGE = 10000;
+    uint256 internal constant MAX_ERROR_COUNT = 7;
+
+    uint256 public offersCounter = 0;
+
+    mapping(uint256 => Offer) public swapOffers;
+
+    mapping(address => uint256) public referralFees;
+
+    /**
+     * @notice Double mapping of signers to nonce groups to nonce states
+     * @dev The nonce group is computed as nonce / 256, so each group of 256 sequential nonces uses the same key
+     * @dev The nonce states are encoded as 256 bits, for each nonce in the group 0 means available and 1 means used
+     */
+    mapping(address => mapping(uint256 => uint256)) internal _nonceGroups;
+
+    /************************************************
+     *  CONSTRUCTOR
+     ***********************************************/
+
+    constructor() {
+        uint256 currentChainId = getChainId();
+        DOMAIN_CHAIN_ID = currentChainId;
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                DOMAIN_TYPEHASH,
+                DOMAIN_NAME,
+                DOMAIN_VERSION,
+                currentChainId,
+                this
+            )
+        );
+    }
+
+    /************************************************
+     *  SETTER
+     ***********************************************/
+
+    /**
+     * @notice Sets the referral fee for a specific referrer
+     * @param referrer is the address of the referrer
+     * @param fee is the fee in percent in 2 decimals
+     */
+    function setFee(address referrer, uint256 fee) external onlyOwner {
+        require(referrer != address(0), "!referrer");
```

Comment:
Can we write a more detailed error message? This is because we are not constrained on bytecode size. This applies to all other error messages too